src/Output.hs

```diff
@@ -1,38 +1,56 @@
+-- In order to add the String type to the Showable typeclass, we need this
+-- pragma. String is an alias for a list of characters, and the default compiler
+-- doesn't let that be in a typeclass because not all its arguments are type
+-- variables.
+{-# LANGUAGE FlexibleInstances #-}
+
 module Output (
-  Showable
-, toLatex
-, toHtml
-, OutputType(..)
-, output
+  OutputType(..)
+, Commentary(..)
+, Showable(..)
+, (.+)
 , Punct(..)
-, Commentary
 ) where
 
-class Showable a where
-    toLatex :: a -> String
-    toHtml :: a -> String
-    toHtml = undefined -- TODO: remove this later
+import Data.Function((&))
+import Data.List.Utils(join)
 
 
 data OutputType = LaTeX
                 | Html
 
-class Outputtable a where
-    outputWrap :: OutputType -> a -> String
 
+newtype Commentary = Commentary [OutputType -> String]
 
--- typeclass instances need to be a specific type constructor with arbitrary
--- type variables passed in. This is such a type.
-newtype Wrap a = Wrap {unwrap :: a}
+instance Showable Commentary where
+    output o (Commentary c) = join "" . map (o &) $ c
+    toCommentary = id
```

Comment:
Mutually recursive types: `Commentary` is in the `Showable` class, and the `Showable` class uses `Commentary` in its definition. One ordering on these definitions isn't obviously better than the other.

Reply:
...no! Define `Commentary`, then define `Showable`, and _then_ specify that `Commentary` is a `Showable` type.